### ISSUE TITLE
Preprocessing does not progress and ends with OOM error

### ISSUE THREAD

**Issue by AlexanderGbd** [autrainer/autrainer#27]
Hi, 
When I want to use `autrainer preprocess` in order to obtain features previous to training I have the following problem.

For a long time it looks like this:

```
Preprocessing datasets...
 - BIRDNETV2-32k

BIRDNETV2-32k:   0%|          | 0/92741 [00:00<?, ?it/s]
```

And then after some time it ends up in an OOM error:
`slurmstepd: error: Detected 1 oom_kill event in StepId=68059.batch. Some of the step tasks have been OOM Killed.`

I already increased the gpu memory and descreased the batch size, but it's still not working which makes me wonder what I am doing wrong?

I use this dataset conf and my data splits are train, dev, test as required:
```
id: BIRDNETV2-32k
_target_: birdnetv2.BIRDNETV2

path: /nas/staff/data_work/AG/HearTheSpecies/birdnet_v2/splits
features_subdir: log_mel_32k_adapted
index_column: file
target_column: species
file_type: npy
file_handler: autrainer.datasets.utils.NumpyFileHandler

criterion: autrainer.criterions.BalancedCrossEntropyLoss
metrics:
  - autrainer.metrics.Accuracy
  - autrainer.metrics.UAR
  - autrainer.metrics.F1
tracking_metric: autrainer.metrics.Accuracy

transform:
  type: grayscale
```

My dataset class simply extends `BaseMLClassificationDataset` and does not change any of the implemented functions, as it should work with the default settings.
Do you have an idea how I could solve this?

**Comment by ramppdev** [comment]
> I already increased the gpu memory and descreased the batch size, but it's still not working which makes me wonder what I am doing wrong?

During preprocessing, we keep all data on the CPU at the moment (this also holds true for any type of transform; unless explicitly changed by the user in the `AbstractDataset` implementation), so no GPU memory should be allocated.
The `batch_size` argument in the main configuration is also not used during preprocessing.

My guess would be that us setting `-n / --num-workers` to `-1` (e.g. trying to use `os.cpu_count()` as a sane default) should be reconsidered and be `1` by default.
It could be possible that `os.cpu_count()` reports all resources available on a node and not just the allocated CPU cores through Slurm.
I will investigate this soon, and in general default to `1` as it is generally more save, thanks for opening the issue!

In the mean time, please let me know if running with a single thread fixes the OOM kill:
```bash
autrainer preprocess -n 1
```

**Comment by ATriantafyllopoulos** [comment]
Can you add a pointer to the code? I've only found [this here](https://github.com/autrainer/autrainer/blob/260ccf9a3ed1cfa90c524f5c043a89394cfcd8b3/autrainer/core/scripts/preprocess_script.py#L184), which seems like it's storing each feature to disk? Why do we keep them all to memory anyway? That won't work for big datasets..

**Comment by AlexanderGbd** [comment]
> In the mean time, please let me know if running with a single thread fixes the OOM kill:
> 
> ```shell
> autrainer preprocess -n 1
> ```
> 
I tried it like this and let it run for a while now (45 minutes), but now it does not even show the progress bar anymore which was the case before at this time. Maybe as it takes longer now, but still, after 45minutes I thought it should be there right?

> Edit: You probably also have to manually delete the `log_mel_32k_adapted` directory, as the preprocessing pipeline would be skipped otherwise and assume that the features are already extracted (pretty sure it exists since the OOM would kill the process after the creation).

It seems it does not even create this folder. At least I couldn't find it.

**Comment by ramppdev** [comment]
This is where we select the number of workers (threads):
https://github.com/autrainer/autrainer/blob/260ccf9a3ed1cfa90c524f5c043a89394cfcd8b3/autrainer/core/scripts/preprocess_script.py#L139-L142

> I've only found [this here](https://github.com/autrainer/autrainer/blob/260ccf9a3ed1cfa90c524f5c043a89394cfcd8b3/autrainer/core/scripts/preprocess_script.py#L184), which seems like it's storing each feature to disk? Why do we keep them all to memory anyway?

We load each file separately here, i assumed this would be automatically collected as `data` is overridden in each iteration, however we can explicitly remove it after saving with `del data`. But i dont think it would hold the entire dataset.

**Comment by AlexanderGbd** [comment]
Okay I checked an older job where I specified a larger memory (32GB) and smaller batch size (16) and after some hours it actually started to process now:

```
Preprocessing datasets...
 - BIRDNETV2-32k

BIRDNETV2-32k:   0%|          | 0/92741 [00:00<?, ?it/s]
BIRDNETV2-32k:   0%|          | 100/92741 [2:35:59<2408:31:05, 93.59s/it]
BIRDNETV2-32k:   0%|          | 200/92741 [2:36:44<997:31:50, 38.81s/it] 
BIRDNETV2-32k:   0%|          | 300/92741 [2:36:58<543:07:04, 21.15s/it]
BIRDNETV2-32k:   0%|          | 400/92741 [2:36:59<328:24:03, 12.80s/it]
BIRDNETV2-32k:   0%|          | 400/92741 [2:37:11<328:24:03, 12.80s/it]
BIRDNETV2-32k:   1%|          | 500/92741 [2:37:28<212:25:46,  8.29s/it]
BIRDNETV2-32k:   1%|          | 600/92741 [2:37:54<142:16:24,  5.56s/it]
BIRDNETV2-32k:   1%|          | 700/92741 [2:37:59<96:08:20,  3.76s/it] 
BIRDNETV2-32k:   1%|          | 700/92741 [2:38:10<96:08:20,  3.76s/it]
```

So it seems now it's working without OOM error.  I'll keep you updated.

**Comment by ramppdev** [comment]
> I tried it like this and let it run for a while now (45 minutes), but now it does not even show the progress bar anymore which was the case before at this time. Maybe as it takes longer now, but still, after 45minutes I thought it should be there right?

Yes it will take longer, but it is definitely strange that no progress bar is showing at all. As a last resort, maybe try to increase the threads slightly, and update the progress bar after each iteration:
```bash
autrainer preprocess -n 4 -p 1
```

> It seems it does not even create this folder. At least I couldn't find it.

This is also strange, as all (sub) directories should be created before extracting the features and after all `csv` files present in the root directory of the dataset are read in to identify all unique files.

**Comment by AlexanderGbd** [comment]
> Yes it will take longer, but it is definitely strange that no progress bar is showing at all. As a last resort, maybe try to increase the threads slightly, and update the progress bar after each iteration:
> 
> ```shell
> autrainer preprocess -n 4 -p 1
> ```
Okay I'll try this for the newest run.

> 
> > It seems it does not even create this folder. At least I couldn't find it.
> 
> This is also strange, as all (sub) directories should be created before extracting the features and after all `csv` files present in the root directory of the dataset are read in to identify all unique files.
> 
I thought so, too. But even now that the one extraction seems to run, there is still no `log_mel_32k_adapted` folder. The folder should be created in the specified `path` of the dataset yaml right?

> Edit:
> 
> > Okay I checked an older job where I specified a larger memory (32GB) and smaller batch size (16) and after some hours it actually started to process now:
> 
> is this using `autrainer preprocess -n 4 -p 1`?

No this is actually still using the old settings where I just ran `autrainer preprocess`.

**Comment by ramppdev** [comment]
> The folder should be created in the specified `path` of the dataset yaml right?

Very interesting, yes the final path should be `{path}/{features_subdir}/some_file.npy` for some extracted file. Both variables come from the dataset configuration YAML file.

**Comment by ramppdev** [comment]
Summary:
- set `-n` and `-p` to `1` by default, and add additional information on how exactly samples are process during `autrainer preprocess` (sequentially, one by one, optionally with threads (`num_workers > 1`) to avoid I/O bottlenecks)
- explicitly delete the processed data after it is saved (`del data`) to decrease memory usage

**Comment by millinma** [comment]
> Okay I checked an older job where I specified a larger memory (32GB) and smaller batch size (16) and after some hours it actually started to process now:

Also weird that it takes "some hours" to arrive there or not? For other datasets I've tried this was all rather fast (done in minutes, at most an hour)?

**Comment by ramppdev** [comment]
> Also weird that it takes "some hours" to arrive there or not? For other datasets I've tried this was all rather fast (done in minutes, at most an hour)?

This was most likely the reason of too many threads and inefficient code. I will open a PR fixing both.
It may still take some time to process data if e.g. data is stored on a NAS.
